### ISSUE TITLE
Hangs on creating Capture or grabbing from offline IP-camera

### ISSUE THREAD

**Issue by cbagpipe** [emgucv/emgucv#22]
Very simple code to reproduce subject. It completely stucks. I can't even kill thread or cancel task whenever it hangs.

```c#
using Emgu.CV;
using System;

namespace OpenCVTest
{
    class Program
    {
        static void Main(string[] args)
        {
            // if camera offline from start, program stucks here.
            var capture = new Capture(@"rtsp://192.168.123.123/");

            // if camera becomes and stays offline mone than 20-30 seconds, program stucks here
            while (capture.Grab())
            {
                Console.Write("grab.");
            }
        }
    }
}
```

**Comment by kolomietsvv** [comment]
If still relevant. I have Inheritd the Capture class (v3.1.0.1 the Capture class still had not been renamed to VideoCapture, version from nuget), and I have overrided the Grab function. That is what I've got. 
`public class VideoCapture : Capture
    {
        public new event EventHandler ImageGrabbed;

        public event EventHandler GrabFailed;

        public VideoCapture(string inputString) : base(inputString) { }

        [HandleProcessCorruptedStateExceptions]
        public override bool Grab()
        {
            try
            {
                if (_ptr == IntPtr.Zero)
                    return false;

                bool grabbed = cveVideoCaptureGrab(Ptr);

                if (grabbed)
                    ImageGrabbed?.Invoke(this, new EventArgs());
                else
                    GrabFailed?.Invoke(this, new EventArgs());
                return grabbed;
            }
            catch
            {
                Thread.Yield();
                GrabFailed?.Invoke(this, new EventArgs());
                return false;
            }
        }

        /// <summary>
        /// Grab a frame
        /// </summary>
        /// <param name="capture">Video capturing structure</param>
        /// <returns>True on success</returns>
        [DllImport("cvextern", CallingConvention = CvInvoke.CvCallingConvention)]
        [return: MarshalAs(CvInvoke.BoolToIntMarshalType)]
        internal static extern bool cveVideoCaptureGrab(IntPtr capture);`

Sometimes _capture.GrabFailed += ReconnectCamera works, but sometimes it steps into grub, but doesn't reach the GrabFailed?.Invoke(this, new EventArgs());. And stays somwhere in bool grabbed = cveVideoCaptureGrab(Ptr);

**Comment by LingboTang** [comment]
I don't think this problem has been fixed up till now, when I read the source code, there's no exception handling for `Grab`, `Retrieve` and `QueryFrame`. I'm using the `*.dll` directly, so I can't simply override that one function. It is really inconvenient, please consider fix it right away.

**Comment by LingboTang** [comment]
After we get an decoding error like:

> h264: cabac decode of qscale diff failed at error decoding

This line:

 `bool grabbed = CvInvoke.cveVideoCaptureGrab(Ptr);`

will always return false because the `Ptr` will point to null, so `cveVideoCaptureGrab` will not grab any frames successfully. How I fix this problem right now is every time the decoding error pops up, I'll just restart the capture. But I think EmguCV should handle this problem on your side.

**Comment by emgucv** [comment]
This is a known FFMPEG/OpenCV behavior: by default FFMPEG has no timeout when connecting to an RTSP stream, so if the camera is unreachable the constructor (or the first `Grab()`) blocks indefinitely.

**Fix 1 — set FFMPEG timeouts before creating VideoCapture**

Set the `OPENCV_FFMPEG_CAPTURE_OPTIONS` environment variable before constructing the capture. Values are in microseconds:

```csharp
// 5-second connect + I/O timeout (5,000,000 µs)
Environment.SetEnvironmentVariable(
    "OPENCV_FFMPEG_CAPTURE_OPTIONS",
    "timeout;5000000|stimeout;5000000");

using var capture = new VideoCapture("rtsp://192.168.123.123/", VideoCapture.API.Ffmpeg);
```

**Fix 2 — use `CapProp.OpenTimeoutMsec` / `CapProp.ReadTimeoutMsec`**

Emgu CV 5 exposes `CapProp.OpenTimeoutMsec = 53` and `CapProp.ReadTimeoutMsec = 54` (added in OpenCV 4.5). Set them immediately after construction for the read timeout:

```csharp
using var capture = new VideoCapture("rtsp://...", VideoCapture.API.Ffmpeg);
capture.Set(CapProp.ReadTimeoutMsec, 5000);
```

**Fix 3 — run the open in a background thread with a cancellation timeout**

Because the constructor itself can hang, wrap it in a `Task` with a timeout:

```csharp
VideoCapture capture = null;
var task = Task.Run(() => capture = new VideoCapture("rtsp://...", VideoCapture.API.Ffmpeg));
if (!task.Wait(TimeSpan.FromSeconds(10)))
    throw new TimeoutException("RTSP open timed out");
```

**Fix 4 — recover from decode errors**

When the stream degrades mid-session, `Grab()` returns `false`. Recreate the capture at that point rather than looping on a dead handle:

```csharp
while (!cancellationToken.IsCancellationRequested)
{
    if (!capture.Grab())
    {
        capture.Dispose();
        capture = new VideoCapture("rtsp://...", VideoCapture.API.Ffmpeg);
        continue;
    }
    // ...
}
```

Closing now.
